### PR TITLE
Allow obtaining &mut OsStr

### PR DESCRIPTION
**Pull request by TyPR124**
```rust
impl DerefMut for OsString {...}              // type Target = OsStr
impl IndexMut<RangeFull> for OsString {...}   // type Output = OsStr
```

---

This change is pulled out of #69937 per @dtolnay 

This implements `DerefMut for OsString` to allow obtaining a `&mut OsStr`. This also implements `IndexMut for OsString`, which is used by `DerefMut`. This pattern is the same as is used by `Deref`.

This is necessary to for methods like `make_ascii_lowercase` which need to mutate the underlying value.

### CODE REVIEW

**Comment by dtolnay** [comment]
@rust-lang/libs: This PR adds **stable** impls for:

```rust
impl DerefMut for OsString {...}              // type Target = OsStr
impl IndexMut<RangeFull> for OsString {...}   // type Output = OsStr
```

The behavior matches the existing `Deref` and `Index` impls for OsString. Currently a &amp;mut OsStr is not very useful but this change will allow introducing (unstable, in a separate PR) things like `make_ascii_uppercase(&mut self)` on OsStr and having it be callable on OsString through derefmut.

@rfcbot fcp merge

File: src/libstd/ffi/os_str.rs [hunk 1]
```diff
@@ -512,6 +528,11 @@ impl OsStr {
         unsafe { &*(inner as *const Slice as *const OsStr) }
     }
 
+    #[inline]
+    fn from_inner_mut(inner: &mut Slice) -> &mut OsStr {
+        unsafe { &mut *(inner as *mut Slice as *mut OsStr) }
```
Comment (Centril) [inline src/libstd/ffi/os_str.rs]: Can you add some comments re. the unsafety here and in the cases below (other files in diff)?

Comment (TyPR124) [reply src/libstd/ffi/os_str.rs]: I can. The unsafety here is essentially the same as the unsafety on `from_inner` and related. Should I add comments for both `from_inner` and `from_inner_mut` ?

Comment (TyPR124) [reply src/libstd/ffi/os_str.rs]: I added some comments to both `from_inner` and `from_inner_mut` as well as related methods.

Comment (Centril) [reply src/libstd/ffi/os_str.rs]: Thanks!

File: src/libstd/sys_common/os_str_bytes.rs [hunk 2]
```diff
@@ -106,9 +106,20 @@ impl Buf {
...
     #[inline]
     pub fn as_slice(&self) -> &Slice {
+        // Safety: Slice just wraps [u8],
+        // and &*self.inner is &[u8], therefore
+        // transmuting &[u8] to &Slice is safe.
         unsafe { mem::transmute(&*self.inner) }
     }
 
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut Slice {
```
Comment (programmerjake) [inline src/libstd/sys_common/os_str_bytes.rs]: One additional safety concern for the internal API: isn't `OsStr` not allowed to contain NULs? assuming that's the case, getting a `&mut [u8]` to it's internals should be `unsafe`.

Comment (CryZe) [reply src/libstd/sys_common/os_str_bytes.rs]: OsStr is a superset of UTF-8, so I don't see how this can be true.

Comment (programmerjake) [reply src/libstd/sys_common/os_str_bytes.rs]: Ah, ok. nevermind.

Comment (BurntSushi) [reply src/libstd/sys_common/os_str_bytes.rs]: Are there any invariants on maintaining valid WTF-8 on Windows though?

Comment (TyPR124) [reply src/libstd/sys_common/os_str_bytes.rs]: @BurntSushi On Windows, `Slice` wraps `Wtf8` and, to the best of my understanding, no methods exist on `Slice` that could invalidate the underlying Wtf8. `Slice` does declare the inner `Wtf8` public, but the `Wtf8` does NOT declare the inner `[u8]` public. So I don't see any way that Wtf8 invariant could be broken. I'll add a couple of comments about upholding any platform-specific encoding, however.

The goal of this is to enable `make_ascii_lowrcase` and `make_ascii_uppercase` for `OsStr`. It is less clear to me that such methods wouldn't break Wtf8, so certainly if there is a problem there please point it out!

Comment (BurntSushi) [reply src/libstd/sys_common/os_str_bytes.rs]: Comments sound great, thanks! (I'm not too familiar with this code, so I was mostly shooting from the hip. :-))

**Pull request merged**